### PR TITLE
feat: create function support lambda

### LINKED ISSUE #14088: feature: create function support lambda expression
**Summary**

```sql
databend-local:) CREATE function my_filter as (list) -> array_filter(list, x -> x > 42);
SyntaxException. Code: 1005, Text = Parameters are not declared: {"x"}.
```

### PR DESCRIPTION
**Pull request by guojidan**
I hereby agree to the terms of the CLA available at: https://docs.databend.com/dev/policies/cla/

## Summary

_Briefly describe what this PR aims to solve. Include background context that will help reviewers understand the purpose of the PR._

Fixes #14088 

## Tests

- [x] Unit Test
- [x] Logic Test
- [ ] Benchmark Test
- [ ] No Test  - _Explain why_

## Type of change

- [ ] Bug Fix (non-breaking change which fixes an issue)
- [x] New Feature (non-breaking change which adds functionality)
- [ ] Breaking Change (fix or feature that could cause existing functionality not to work as expected)
- [ ] Documentation Update
- [ ] Refactoring
- [ ] Performance Improvement
- [ ] Other (please describe):

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/databend/14104)

### CODE REVIEW

File: src/query/sql/src/planner/udf_validator.rs [hunk 1]
```diff
@@ -46,7 +47,8 @@ impl UDFValidator {
             return Err(ErrorCode::SyntaxException("Recursive UDF is not supported"));
         }
         let expr_params = &self.expr_params;
-        let parameters = self.parameters.iter().cloned().collect::<HashSet<_>>();
+        let mut parameters = self.parameters.iter().cloned().collect::<HashSet<_>>();
+        parameters.extend(self.lambda_parameters.iter().cloned().collect::<HashSet<_>>().into_iter());
```
Comment (sundy-li) [inline src/query/sql/src/planner/udf_validator.rs:51]: I'm sure we can avoid another collect in this line.

File: tests/sqllogictests/suites/base/05_ddl/05_0010_ddl_create_udf.test [hunk 2]
```diff
@@ -22,6 +22,18 @@ CREATE FUNCTION IF NOT EXISTS isnotempty_with_desc AS(p) -> not(is_null(p)) DESC
 statement error 2603
 CREATE FUNCTION isnotempty_with_desc AS(p) -> not(is_null(p)) DESC = 'This is a description'
 
+statement ok
+CREATE FUNCTION with_lambda AS(list) -> array_filter(list, x -> x > 42)
```
Comment (sundy-li) [inline tests/sqllogictests/suites/base/05_ddl/05_0010_ddl_create_udf.test:26]: Need tests to validate whether the new function works or not.

Comment (guojidan) [reply tests/sqllogictests/suites/base/05_ddl/05_0010_ddl_create_udf.test:26]: nice, I didn't think of that

File: tests/sqllogictests/suites/base/05_ddl/05_0010_ddl_create_udf.test [hunk 3]
```diff
@@ -22,6 +22,33 @@ CREATE FUNCTION IF NOT EXISTS isnotempty_with_desc AS(p) -> not(is_null(p)) DESC
...
+SELECT with_lambda(arr) from array_int64_table
+----
+   [3,4]
+   [5,6,7,8]
+
+statement error 1005
+CREATE FUNCTION my_filter as (list) -> array_filter(list, x -> y > 2);
+
 statement ok
 DROP FUNCTION isnotempty
```
Comment (sundy-li) [inline tests/sqllogictests/suites/base/05_ddl/05_0010_ddl_create_udf.test:53]: need drop function `with_lambda`

File: tests/sqllogictests/suites/base/05_ddl/05_0010_ddl_create_udf.test [hunk 4]
```diff
@@ -22,6 +22,33 @@ CREATE FUNCTION IF NOT EXISTS isnotempty_with_desc AS(p) -> not(is_null(p)) DESC
...
+statement ok
+CREATE TABLE IF NOT EXISTS array_int64_table(arr ARRAY(INT64))
+
+statement ok
+INSERT INTO array_int64_table VALUES ([1, 2, 3, 4]), ([5, 6, 7, 8])
+
+query T
+SELECT with_lambda(arr) from array_int64_table
+----
+   [3,4]
```
Comment (sundy-li) [inline tests/sqllogictests/suites/base/05_ddl/05_0010_ddl_create_udf.test:46]: incorrect leading spaces

**Review by sundy-li — APPROVED** [review]

**Pull request merged**